src/lib.rs

```diff
@@ -181,41 +181,45 @@ pub mod serde {
 ///
 /// // Encode the Value to bytes:
 /// let mut bytes = Vec::new();
-/// scale_value::scale::encode_as_type(&value, type_id, &registry, &mut bytes).unwrap();
+/// scale_value::scale::encode_as_type(&value, &type_id, &registry, &mut bytes).unwrap();
 ///
 /// // Decode the bytes back into a matching Value.
 /// // This value contains contextual information about which type was used
 /// // to decode each part of it, which we can throw away with `.remove_context()`.
-/// let new_value = scale_value::scale::decode_as_type(&mut &*bytes, type_id, &registry).unwrap();
+/// let new_value = scale_value::scale::decode_as_type(&mut &*bytes, &type_id, &registry).unwrap();
 ///
 /// assert_eq!(value, new_value.remove_context());
 /// ```
 pub mod scale {
     use crate::prelude::*;
     use scale_encode::EncodeAsType;
 
-    pub use crate::scale_impls::{DecodeError, TypeId};
+    pub use crate::scale_impls::DecodeError;
     pub use scale_encode::Error as EncodeError;
     pub use scale_info::PortableRegistry;
+    pub use scale_type_resolver::TypeResolver;
 
     /// Attempt to decode some SCALE encoded bytes into a value, by providing a pointer
     /// to the bytes (which will be moved forwards as bytes are used in the decoding),
     /// a type ID, and a type registry from which we'll look up the relevant type information.
-    pub fn decode_as_type(
+    pub fn decode_as_type<R: TypeResolver>(
         data: &mut &[u8],
-        ty_id: TypeId,
-        types: &PortableRegistry,
-    ) -> Result<crate::Value<TypeId>, DecodeError> {
+        ty_id: &R::TypeId,
+        types: &R,
+    ) -> Result<crate::Value<R::TypeId>, DecodeError>
+    where
+        R::TypeId: Clone,
```

Comment:
nit, I would prefer some consistency here w.r.t to trait bounds i.e, either specify the bounds on the generic type or in a where clause.

Such as 

```rust
   pub fn decode_as_type<R>(
        data: &mut &[u8],
        ty_id: TypeId,
        types: &PortableRegistry,
    ) -> Result<crate::Value<TypeId>, DecodeError> {
        ty_id: &R::TypeId,
        types: &R,
    ) -> Result<crate::Value<R::TypeId>, DecodeError>
    where
        R: TypeResolver,
        R::TypeId: Clone,
```

Reply:
I'd normally go for consistency and would suggest them both to be in the `where` clause (I think I made this tweak actually and didn't think about it at the time!)